include/pybind11/pytypes.h

```diff
@@ -390,6 +390,13 @@ inline void setattr(handle obj, handle name, handle value) {
 inline void setattr(handle obj, const char *name, handle value) {
     if (PyObject_SetAttrString(obj.ptr(), name, value.ptr()) != 0) { throw error_already_set(); }
 }
+
+inline long hash(object obj) {
```

Comment:
`handle` is generally used for function parameters to avoid an extra incref/decref of copying an `object`.

`Py_hash_t` is needed on Python 3 to resolve the compiler warnings, but it's not available on Python 2. Looking at the docs, `ssize_t` should satisfy the return type of both 2 and 3:

> Changed in version 3.2: The return type is now Py_hash_t. This is a signed integer the same size as Py_ssize_t.

Reply:
Thanks. Not sure how that ended up being `object` - it was `handle` in my brain. I decided to deal with the return type by letting `decltype` handle it.

Reply:
For the return type: It would be nice for pybind11's API to have the same type for both Python 2 and 3, so that we don't pass on these inconsistencies to users (unless it's absolutely necessary). If `ssize_t` is compatible with both, it would be preferable IMO.

Reply:
I don't think anything in the standard that prevents `long` from being wider than `Py_ssize_t`, but I don't know of any contemporary systems where that happens (it would probably have been true on 16-bit systems). I've switched to Py_ssize_t and added a `static_assert` just to be safe.